src/desktop.c

```diff
@@ -170,42 +139,30 @@ desktop_cycle_view(struct server *server, struct view *start_view,
 	/* Make sure to have all nodes in their actual ordering */
 	osd_preview_restore(server);
 
-	if (!start_view) {
-		start_view = first_view(server);
-		if (!start_view || start_view != server->active_view) {
-			return start_view;  /* may be NULL */
-		}
-	}
-	struct view *view = start_view;
-	struct wlr_scene_node *node = &view->scene_tree->node;
-
-	assert(node->parent);
-	struct wl_list *list_head = &node->parent->children;
-	struct wl_list *list_item = &node->link;
-	struct wl_list *(*iter)(struct wl_list *list);
-
-	/* Scene nodes are ordered like last node == displayed topmost */
-	iter = dir == LAB_CYCLE_DIR_FORWARD ? get_prev_item : get_next_item;
-
-	do {
-		list_item = iter(list_item);
-		if (list_item == list_head) {
-			/* Start / End of list reached. Roll over */
-			list_item = iter(list_item);
-		}
-		node = wl_container_of(list_item, node, link);
-		if (!node->data) {
-			/* We found some non-view, most likely the region overlay */
-			view = NULL;
-			continue;
-		}
-		view = node_view_from_node(node);
+	struct view *(*iter)(struct wl_list *head, struct view *view,
+		enum lab_view_criteria criteria);
+	bool forwards = dir == LAB_CYCLE_DIR_FORWARD;
+	iter = forwards ? view_next_no_head_stop : view_prev_no_head_stop;
 
-		enum property skip = window_rules_get_property(view, "skipWindowSwitcher");
-		if (view_is_focusable(view) && skip != LAB_PROP_TRUE) {
+	/*
+	 * TODO: These criteria are the same as in display_osd() in osd.c
+	 * for the time being.
+	 *
+	 * A future improvement could be to make this configurable for example
+	 * in rc.xml and then use rc.cycle_view_criteria (or whatever) both
+	 * here and in the osd.c window-switcher code
+	 */
+	enum lab_view_criteria criteria = LAB_VIEW_CRITERIA_CURRENT_WORKSPACE
+		| LAB_VIEW_CRITERIA_NO_SKIP_WINDOW_SWITCHER;
+
+	/* On the first alt-tab, there is no start_view to we have to set it */
+	struct view *view = start_view ? : iter(&server->views, NULL, criteria);
+
+	for_each_view_from_bidirectional(view, &server->views, iter, criteria) {
+		if (view_is_focusable(view)) {
```

Comment:
I think we do check `view_is_focusable()` in `matches_criteria()`.

Reply:
Thanks for spotting that. It actually simplifies things ever more because all we need to do it call `iter()`. No more `for_each_view_from_bidirectional` :smile: